include/deal.II/base/mpi.h

```diff
@@ -708,6 +709,416 @@ namespace Utilities
+      bool
+      check_own_state();
+
+      /**
+       * Signal to all other ranks that this rank has received all request
+       * answers via entering IBarrier.
+       */
+      void
+      signal_finish();
+
+      /**
+       * Check if all ranks have received all their request answers, i.e.
+       * all ranks have reached the IBarrier.
+       */
+      bool
+      check_global_state();
+
+      /**
+       * A request message from another rank has been received: process the
+       * request and send an answer.
+       */
+      void
+      process_requests();
+
+      /**
+       * Start to send all requests via ISend and post IRecvs for the incoming
+       * answer messages.
+       */
+      void
+      start_communication();
+
+      /**
+       * After all rank has received all answers, the MPI data structures can be
+       * freed and the received answers can be processed.
+       */
+      void
+      clean_up_and_end_communication();
+    };
+
+    /**
+     * This class implements ConsensusAlgorithm, using a two step approach. In
+     * the first step the source ranks are determined and in the second step
+     * a static sparse data exchange is performed.
+     *
+     * @note In contrast to ConsensusAlgorithm_NBX, this class splits the same
+     *       task into two distinct steps. In the first step, all processes are
+     *       identified who wants to send a request to this process. In the
+     *       second step, the data is exchanged. However, since - in the second
+     *       step - now it is clear how many requests have to be answered, i.e.
+     *       when this process can stop waiting for requests, no IBarrier is
+     *       needed.
+     *
+     * @note The function compute_point_to_point_communication_pattern is used
+     *       to determine the source processes, which implements an
+     *       PEX-algorithm from Hoefner et. al. "Scalable Communication
+     *       Protocols for Dynamic Sparse Data Exchange"
+     *
+     * @author Peter Munch, 2019
+     */
+    template <typename T1, typename T2>
+    class ConsensusAlgorithm_PEX : public ConsensusAlgorithm<T1, T2>
+    {
+    public:
+      // Unique tags to be used during Isend and Irecv
+      static const unsigned int tag_request  = 14;
+      static const unsigned int tag_delivery = 15;
+
+      /**
+       * Constructor.
+       *
+       * @param process Process to be run during consensus algorithm.
+       * @param comm MPI Communicator
+       */
+      ConsensusAlgorithm_PEX(ConsensusAlgorithmProcess<T1, T2> &process,
+                             const MPI_Comm &                   comm);
+
+      /**
+       * Destructor.
+       */
+      virtual ~ConsensusAlgorithm_PEX();
+
+      /**
+       * Run consensus algorithm.
+       */
+      void
+      run() override;
+
+    private:
+#ifdef DEAL_II_WITH_MPI
+      // list of ranks of processes this processes wants to send a request to
+      std::vector<unsigned int> targets;
```

Comment:
Please convert comments to doxygen style (and a separate comment for each variable in the list below).